File: src/main/java/me/rabbittv/terraTrees/utils/SpawnStructure.java
package me.rabbittv.terraTrees.utils;

import com.dfsek.terra.api.config.ConfigPack;
import com.dfsek.terra.api.registry.key.RegistryKey;
import com.dfsek.terra.api.structure.Structure;
import com.dfsek.terra.api.util.Rotation;
import com.dfsek.terra.api.util.vector.Vector3Int;
import com.dfsek.terra.bukkit.world.BukkitServerWorld;
import me.rabbittv.terraTrees.TerraTrees;
import org.bukkit.Location;
import org.bukkit.World;
import java.util.Optional;
import java.util.Random;

public class SpawnStructure {

    private final TerraTrees plugin;

    public SpawnStructure(TerraTrees plugin) {
        this.plugin = plugin;
    }

    public void spawnTerraStructure(World world, Location location, String structureId) {
        BukkitServerWorld TerraWorld = new BukkitServerWorld(world);
        Vector3Int vector = Vector3Int.of((int)Math.floor(location.getX()), (int)Math.floor(location.getY()), (int)Math.floor(location.getZ()));

        ConfigPack pack = TerraWorld.getPack();
        if (pack == null) {
            plugin.getLogger().warning("Could not find a config pack for the current world.");
        } else {
            RegistryKey structureKey = RegistryKey.parse(structureId);
            Optional<Structure> structureOptional = pack.getRegistry(Structure.class).get(structureKey);
            if (structureOptional.isEmpty()) {
                plugin.getLogger().warning("Structure with ID '" + structureId + "' not found in config pack.");
            }
            Structure structure = structureOptional.get();

            Random random = new Random();
            Rotation rotation = Rotation.NONE;

            boolean generated = structure.generate(vector, TerraWorld, random, rotation);

            if (generated && plugin.settings.getBoolean("debug")) {
                plugin.getLogger().info("Successfully spawned structure '" + structureId + "' at " + location.toString());
            } else if (plugin.settings.getBoolean("debug")) {
                plugin.getLogger().warning("Failed to spawn structure '" + structureId + "' at " + location.toString());
            }
        }
    }

}

File: src/main/kotlin/me/rabbittv/terraTrees/utils/SpawnStructure.kt
package me.rabbittv.terraTrees.utils

import com.dfsek.terra.api.registry.key.RegistryKey
import com.dfsek.terra.api.structure.Structure
import com.dfsek.terra.api.util.Rotation
import com.dfsek.terra.api.util.vector.Vector3Int
import com.dfsek.terra.bukkit.world.BukkitServerWorld
import me.rabbittv.terraTrees.TerraTrees
import org.bukkit.Location
import org.bukkit.World
import java.util.Random

class SpawnStructure(private val plugin: TerraTrees) {


    fun spawnTerraStructure(world: World, location: Location, StructureID: String?) {
        val TerraWorld = BukkitServerWorld(world)
        val vector = Vector3Int.of(location.blockX, location.blockY, location.blockZ)

        val TerraPack = TerraWorld.pack
        if (TerraPack == null) {
            plugin.logger.warning("The current world does not have a Terra pack")
        } else {
            val StructureKey = RegistryKey.parse(StructureID)
            val StructureOptional = TerraPack.getRegistry(Structure::class.java).get(StructureKey)
            if (StructureOptional.isEmpty) {
                plugin.logger.warning("Structure $StructureID not found.")
            }
            val structure = StructureOptional.get()
            val random = Random()
            val rotation = Rotation.NONE
            val generated = structure.generate(vector, TerraWorld, random, rotation)

            if (generated && plugin.settings.getBoolean("debug")){
                plugin.logger.info("Successfully spawned structure $StructureID at ${location.blockX}, ${location.blockY}, ${location.blockZ}")
            } else if (plugin.settings.getBoolean("debug")) {
                plugin.logger.warning("Failed to spawn structure $StructureID at ${location.blockX}, ${location.blockY}, ${location.blockZ}.")
            }
        }
    }

}